nemo/collections/asr/models/configs/ctc_models_config.py

```diff
@@ -0,0 +1,173 @@
+
+from nemo.core.config import modelPT as model_cfg
+
+
+@dataclass
+class BaseASRPreprocessorConfig:
+    pass
+
+
+@dataclass
+class AudioToMelSpectrogramPreprocessorConfig(BaseASRPreprocessorConfig):
+    _target_: str = "nemo.collections.asr.modules.AudioToMelSpectrogramPreprocessor"
+    sample_rate: int = 16000
+    window_size: float = 0.02
+    window_stride: float = 0.01
+    n_window_size: Optional[int] = None
+    n_window_stride: Optional[int] = None
+    window: str = "hann"
+    normalize: str = "per_feature"
+    n_fft: Optional[int] = None
+    preemph: float = 0.97
+    features: int = 64
+    lowfreq: int = 0
+    highfreq: Optional[int] = None
+    log: bool = True
+    log_zero_guard_type: str = "add"
+    log_zero_guard_value: float = 2 ** -24
+    dither: float = 1e-5
+    pad_to: int = 16
+    frame_splicing: int = 1
+    stft_exact_pad: bool = False
+    stft_conv: bool = False
+    pad_value: int = 0
+    mag_power: float = 2.0
+
+
+@dataclass
+class SpecAugmentConfig:
+    _target_: str = "nemo.collections.asr.modules.SpectrogramAugmentation"
+    freq_masks: int = 0
+    time_masks: int = 0
+    freq_width: int = 0
+    time_width: Optional[Any] = 0
+    rect_masks: int = 0
+    rect_time: int = 0
+    rect_freq: int = 0
+
+
+@dataclass
+class EncDecCTCDatasetConfig(model_cfg.DatasetConfig):
+    manifest_filepath: Optional[str] = MISSING
+    sample_rate: int = MISSING
+    labels: List[str] = MISSING
+    trim_silence: bool = False
+
+    # Tarred dataset support
+    is_tarred: bool = False
+    tarred_audio_filepaths: Optional[str] = None
+    shuffle_n: int = 0
+
+    # Optional
+    int_values: Optional[int] = None
+    augmentor: Optional[Dict[str, Any]] = None
+    max_duration: Optional[float] = None
+    min_duration: Optional[float] = None
+    max_utts: int = 0
+    blank_index: int = -1
+    unk_index: int = -1
+    normalize: bool = False
+    trim: bool = True
+    load_audio: bool = True
+    parser: Optional[str] = 'en'
+    add_misc: bool = False
+
+
+@dataclass
+class JasperEncoderConfig:
+    filters: int = MISSING
+    repeat: int = MISSING
+    kernel: List[int] = MISSING
+    stride: List[int] = MISSING
+    dilation: List[int] = MISSING
+    dropout: float = MISSING
+    residual: bool = MISSING
+
+    # Optional arguments
+    groups: int = 1
+    separable: bool = False
+    heads: int = -1
+    residual_mode: str = "add"
+    residual_dense: bool = False
+    se: bool = False
+    se_reduction_ratio: int = 8
+    se_context_size: int = -1
+    se_interpolation_mode: str = 'nearest'
+    kernel_size_factor: float = 1.0
+    stride_last: bool = False
+
+
+@dataclass
+class ConvASREncoderConfig:
+    _target_: str = 'nemo.collections.asr.modules.ConvASREncoder'
+    jasper: Optional[JasperEncoderConfig] = field(default_factory=list)
+    activation: str = MISSING
+    feat_in: int = MISSING
+    normalization_mode: str = "batch"
+    residual_mode: str = "add"
```

Comment:
Changing residual mode default here will *not* be reflected in the residual mode at the JasperEncoderConfig level! Code must be changed to support this (can be done easily).

Reply:
Both thumbs up for this - and generally from moving all defaults from models/datasets to the associated configs

Reply:
@tkornuta-nvidia This is actually *not good*. By default, in Citrinet, I dont add residual mode (therefore it defaults to global default mode of "add"), but in certain places, I am required to override "add" with "stride_add" (if I remember correctly). In yaml, this is perfectly fine and very convenient.

However in Structured Configs, I would have to explicitly update not only the global one here, but every single local one within the list manually. Its a dissapointing limitation to not be able to map from a global to a local inside a dataclass.

Reply:
If you have hardcoded override of a parameter in the code, then maybe it shouldn't be a configuration parameter?

I still don't understand what is "not good" in having defaults in Structured Configs...

Reply:
Its useful and required to be overriden from yaml / code. Structured configs should adapt to support our use cases, not the other way around. Thats my opinion on this.